perf/proxy_thr.cpp

```diff
@@ -0,0 +1,450 @@
+    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+    License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "../include/zmq.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <time.h>
+#include <stdarg.h>
+#include <string.h>
+#include <string>
+
+#include "platform.hpp"
+
+#if defined ZMQ_HAVE_WINDOWS
+#include <windows.h>
+#include <process.h>
+#else
+#include <pthread.h>
+#include <unistd.h>
+#endif
+
+
+//
+// Asynchronous proxy benchmark using ZMQ_XPUB_NODROP.
+//
+// Topology:
+//
+//   XPUB                      SUB
+//    |                         |
+//    +-----> XSUB -> XPUB -----/
+//    |       ^^^^^^^^^^^^
+//   XPUB      ZMQ proxy
+//
+// All connections use "inproc" transport. The two XPUB sockets start
+// flooding the proxy. The throughput is computed using the bytes received
+// in the SUB socket.
+//
+
+
+#define HWM 10000
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(x) (sizeof (x) / sizeof (*x))
+#endif
+
+#define TEST_ASSERT_SUCCESS_ERRNO(expr)                                        \
+    test_assert_success_message_errno_helper (expr, NULL, #expr)
+
+
+static uint64_t message_count = 0;
+static size_t message_size = 0;
+
+
+typedef struct
+{
+    void *context;
+    int thread_idx;
+    const char *frontend_endpoint[4];
+    const char *backend_endpoint[4];
+    const char *control_endpoint;
+} proxy_hwm_cfg_t;
+
+
+int test_assert_success_message_errno_helper (int rc_,
+                                              const char *msg_,
+                                              const char *expr_)
+{
+    if (rc_ == -1) {
+        char buffer[512];
+        buffer[sizeof (buffer) - 1] =
+          0; // to ensure defined behavior with VC++ <= 2013
+        printf ("%s failed%s%s%s, errno = %i (%s)", expr_,
+                msg_ ? " (additional info: " : "", msg_ ? msg_ : "",
+                msg_ ? ")" : "", zmq_errno (), zmq_strerror (zmq_errno ()));
+        exit (1);
+    }
+    return rc_;
+}
+
+void recv_string_expect_success (void *socket_, const char *str_, int flags_)
+{
+    const size_t len = str_ ? strlen (str_) : 0;
+    char buffer[255];
+
+    if (len > sizeof (buffer)) {
+        printf ("error: recv_string_expect_success cannot be "
+                "used for strings longer than 255 characters");
+        exit (1);
+    }
+
+    const int rc = TEST_ASSERT_SUCCESS_ERRNO (
+      zmq_recv (socket_, buffer, sizeof (buffer), flags_));
+
+    if ((int) len != rc) {
+        printf ("invalid response length: expected %zu, received %d", len, rc);
+    }
+
+    if (str_) {
+        if (len != strnlen (buffer, sizeof (buffer))) {
+            printf (
+              "invalid response string length: expected %zu, received %zu", len,
```

Comment:
../perf/proxy_thr.cpp:123:78: error: ISO C++ does not support the ‘z’ gnu_printf length modifier [-Werror=format]